File: App/Domain/Statistics.cs
namespace App
{
    using System;
    using System.Collections.Generic;
    using System.Linq;

    public class Statistics
    {
        private readonly List<Activity> _listOfActivities;
        private readonly DateProvider _dateProvider;

        public Statistics(List<Activity> listOfActivities, DateProvider dateProvider)
        {
            _listOfActivities = listOfActivities;
            _dateProvider = dateProvider;
        }

        public int TotalTimeSpan()
        {
            var smallestDate = _listOfActivities.Min(activity => activity.Date);
            var totalDays = _dateProvider.GetCurrentDate().Subtract(smallestDate).TotalDays;

            return Convert.ToInt32(Math.Floor(totalDays));
        }

        public int TotalDurationOfActivities()
        {
            return _listOfActivities.Sum(activity => activity.Duration);
        }

        public int TotalDurationOfActivities(ActivityType activityType)
        {
            return _listOfActivities.Where(activity => activity.ActivityType == activityType).Sum(activity => activity.Duration);
        }

        public int TotalNumberOfMovies
        {
            get { return _listOfActivities.Count(activity => activity.ActivityType == ActivityType.Movie); }
        }

        public int TotalNumberOfSeries
        {
            get { return _listOfActivities.Count(activity => activity.ActivityType == ActivityType.Series); }
        }

        public double AverageIntervalBetweenActivities
        {
            get
            {
                var sortedListOfActivities = _listOfActivities.OrderBy(activity => activity.Date).ToList();

                return CalculateAverageIntervalBetweenSortedActivities(sortedListOfActivities);
            }
        }

        public double AverageIntervalBetweenCinemaVisits
        {
            get
            {
                var sortedListOfActivities = _listOfActivities
                    .Where(activity => activity.WatchedInCinema)
                    .OrderBy(activity => activity.Date).ToList();

                return CalculateAverageIntervalBetweenSortedActivities(sortedListOfActivities);
            }
        }

        private static double CalculateAverageIntervalBetweenSortedActivities(List<Activity> sortedListOfActivities)
        {
            var totalDays = 0.0;

            for (var i = 0; i < sortedListOfActivities.Count - 1; i++)
            {
                var nextActivityDate = sortedListOfActivities[i + 1].Date;
                var previousActivityDate = sortedListOfActivities[i].Date;

                totalDays += nextActivityDate.Subtract(previousActivityDate).TotalDays;
            }

            var numberOfIntervals = sortedListOfActivities.Count - 1;

            return totalDays / numberOfIntervals;
        }
    }
}

File: App/NodeServer/statistics.js
'use strict';
var _ = require('lodash');
var moment = require('moment');

var statistics = function(activities) {
    var unicorn = {};

    unicorn.totalDurationOfActivities = totalDuration();
    unicorn.totalTimeSpan = totalTimeSpan();
    unicorn.totalDurationOfMovies = totalDurationOfMovies();
    unicorn.totalDurationOfSeries = totalDurationOfSeries();
    unicorn.averageIntervalBetweenActivities = averageIntervalBetweenActivities();
    unicorn.averageIntervalBetweenCinemaVisits = averageIntervalBetweenCinemaVisits();
    unicorn.totalNumberOfMovies = totalNumberOfMovies();
    unicorn.totalNumberOfSeries = totalNumberOfSeries();

    function totalDuration() {
        return _.reduce(activities, function(totalDuration, n) {
            return totalDuration + n.duration;
        }, 0);
    }

    function totalTimeSpan() {
        var activityWithMinDate = _.min(activities, function(activity) {
            return new Date(activity.date);
        });

        var minDate = moment(activityWithMinDate.date);
        var now = moment();
        var differenceInDays = now.diff(minDate, 'days');

        return differenceInDays;
    }

    function totalDurationOfMovies() {
        return _.reduce(activities, function(totalDuration, n) {
            return (n.activityType === 'Movie') ? totalDuration + n.duration : totalDuration;
        }, 0);
    }

    function totalDurationOfSeries() {
        return _.reduce(activities, function(totalDuration, n) {
            return (n.activityType === 'Series') ? totalDuration + n.duration : totalDuration;
        }, 0);
    }

    function averageIntervalBetweenActivities() {
        if (activities.length <= 1) {
            return 0;
        }

        var sortedActivities = _.chain(activities)
            .uniq(function(activity) { return activity.date })
            .sortBy(function(activity) { return new Date(activity.date); })
            .value();

        return _calculateInterval(sortedActivities);
    }

    function averageIntervalBetweenCinemaVisits() {
        var sortedActivities = _.chain(activities)
            .uniq(function(activity) { return activity.date })
            .sortBy(function(activity) { return new Date(activity.date); })
            .filter(function(activity) { return activity.watchedInCinema === true })
            .value();

        if (sortedActivities.length <= 1) {
            return 0;
        }

        return _calculateInterval(sortedActivities);
    }

    function _calculateInterval(sortedActivities) {
        var totalDays = 0;
        var numberOfIntervals = sortedActivities.length - 1;

        for (var i = 0; i < numberOfIntervals; i++)
        {
            var nextActivityDate = moment(sortedActivities[i + 1].date);
            var previousActivityDate = moment(sortedActivities[i].date);

            totalDays += nextActivityDate.diff(previousActivityDate, 'days');
        }

        return totalDays / numberOfIntervals;
    }

    function totalNumberOfMovies() {
        return _calculateTotalNumberOfActities('movie');
    }

    function totalNumberOfSeries() {
        return _calculateTotalNumberOfActities('series');
    }

    function _calculateTotalNumberOfActities(type) {
        var activitiesCount = _.filter(activities, function(activity) {
            return activity.activityType.toLowerCase() === type;
        }).length;

        return activitiesCount;
    }

    return unicorn;
};

module.exports = statistics;